### ISSUE TITLE
Design SQL queries to generate Example 5 from mockup

### ISSUE THREAD

**Issue by ultimike** [communityinclusion/escoach#20]
Building off of #8 and #11...

See the "Gitlab #20" tab of [the Google Sheet](https://docs.google.com/spreadsheets/d/13T2JrHmX-xpxbkB5wk5z9Br4V5hauoWlL4gy0pg7ZXA/edit#gid=2066763682).

![Screen Shot 2021-05-07 at 1 58 44 PM](https://user-images.githubusercontent.com/1727889/117490087-5beba300-af3c-11eb-95e7-8c5eaa7ac508.png)

**Comment by ultimike** [comment]
@escoach I'm still a little fuzzy on how to handle things when the time frame is "monthly" or "quarterly" trends, but I took my best shot.  First, here's the query I used:

```
/* SQL queries for Gitlab #20 */
/* answer483 = 11659 is "What = Finding, securing jobs" */
/* answer481 = 11641 is "Who = Family, friends...", answer590 = 12119 is "Who else = Family, friends..." */
/* answer482 = 11651 is "Where = At a business... " */
SELECT MONTH(date_submitted),
       count(*) AS TotalAll,
       sum(case when (((answer481 = 11641) OR (answer590 = 12119)) AND (answer482 = 11651)) then 1 else 0 end) AS SelectedAll,
       sum(case when (provider = 'no provider') then 1 else 0 end) AS TotalProvider,
       sum(case when ((answer481 = 11641) OR (answer590 = 12119) AND (answer482 = 11651) AND (provider = 'no provider')) then 1 else 0 end) AS SelectedProvider,
       sum(case when (email = 'paul.foos@umb.edu') then 1 else 0 end) AS TotalMe,
       sum(case when ((answer481 = 11641) OR (answer590 = 12119) AND (answer482 = 11651) AND (email = 'paul.foos@umb.edu')) then 1 else 0 end) AS SelectedMe
FROM surveycampaign_results
WHERE answer483 = 11659
GROUP BY MONTH(date_submitted);
```

As you can see, I went a little different than the mockup; I selected a "What", a "Who", and a "Where", then grouped the counts by month of the year (there is no data for May or June in the dataset I have.

Take a look at the Gitlab #21 tab and let me know if I'm on the right track. Again, assume the `#DIV/0!` errors will be displayed as `00:00` in the final output.

Once we get this one correct, for the quarterly trends, I'll have to be a bit more creative with the SQL `GROUP BY` bit.

-mike

**Comment by escoach** [comment]
@ultimike I reviewed Gitlab#20 (There is no Gitlab#21. Typo?).
I have two initial comments before I take a closer look at Gitlab#20.

A) I am not entirely sure about what would be the advantage of combining a "What", "Who", and "Where". From our perspective, it's a concern. Rationale: it reduces substantially the 'n', especially at the "MyData" level. And especially given that we are looking at one month worth of data only. Note that each user can submit at most about 20 surveys per month (one per workday). So there are going to be very few instances of a user reporting "Finding jobs" while interacting with "Families", and while in a "Business".  That is why the mockup Example 5 says that the user can pick either WHO or Where, not both. 

In fact, the more I think about it, the more I am getting nervous about whether we can even allow picking either Who or Where. In the spirit of keeping things simple, would it be better to use only the WHAT data (any combination of one or more of the What items) for the trend data? Once that is done, we can come back on whether to add the "Either who or Where" query? Still in the interest of keeping things simple, would it be useful to focus on the quarterly trend first? And then once that is figured out, do the monthly trends? Because of the low numbers, we might end up realizing that the monthly trends do not have enough data to be meaningful. Thoughts? 

B) The total Ns (Row 16) seems too low? For example, I would expect that "TotalAll" (B16) be 327 i.e. the total number of surveys collected since the onset of data collection?

**Comment by ultimike** [comment]
@escoach 

Yes - "21" was a typo - should be "20".

For now, I would say that let's press forward allowing a user to select a single "What" and a single "Who" or a single "Where". If there isn't enough data for it to make sense, then we can easily remove the ability for them to select a "Who" or "Where".

Same thing for quarterly vs. monthly - I don't think we have to make any decisions yet, let's see how the data looks and then we can hide monthly later if need be.

I made a tweak to the query (updated above in my previous comment) - do the results on the Google Sheet look better now?

-mike

**Comment by escoach** [comment]
@ultimike I agree with your approach: see what happens with the data before we make a decision.
Gitlab#20 looks correct now.

**Comment by ultimike** [comment]
I need to add a WHERE clause to only include the last 12 months of data.

**Comment by escoach** [comment]
@ultimike would it be relatively easy to set it up so that if a user picks "Quarterly", then the clause includes the past 3 years? 
In other words, the chart always displays 12 data points: 12 months worth of data for the monthly trends and 3 years (12 quarters) for the quarterly trends.

Our mockup did not make this distinction and it is not a big deal for now. So for the sake of simplicity, I am ok if the clause is set to include 12 months only. We probably can work on that once we have more than one year's worth of new data.

**Comment by ultimike** [comment]
@escoach Yes, we should be able to limit the "monthly trends" data to the last 12 months, while not limiting the quarterly trends at all.

As I was testing the queries for this, it struck me that we're not limiting any of the other (non-monthly/quarterly trends) by date. In other words, all the other examples we've worked through include **all** data, not just the last 12 months. Is this your preference, or should _everything_ except quarterly trends show _only_ data from the last 12 months?

-mike

**Comment by ultimike** [comment]
Example queries for both Monthly and Quarterly trends:

```

**Comment by escoach** [comment]
@ultimike Good observation. In an effort to keep it simple, we did not ask for limiting the "Up to date" data to a certain period of time. If a user is interested in the last 12 months, the user can always select the monthly trend view. I hope that this makes sense. So to answer your question, I think that we are ok. Or let me know if I am oversimplifying. Thank you.

On a parallel issue, right now the dataset includes fake data from our testing. We will need to exclude these data. The easiest thing will be to manually delete those records. But if it is not too complicated, it would be useful to have a switch so we as the administrators can set which period of past data to include/exclude. Not a big deal if this complicates things. We can simply delete the fake data once we are done with testing and start to enroll real users. Please let me know if you have any suggestions.

**Comment by ultimike** [comment]
@escoach Understood - the only data that will be date limited will be the monthly trend.

As for the fake data, I've added #26 (low priority) for adding a switch to ignore fake data, but for now, let's just leave it in there and manually remove it in the future, if necessary. I don't want to lose focus of the main tasks.

-mike

**Comment by escoach** [comment]
@ultimike @thinkcollege Hi Mike, this coming Thursday 06/10 we have a meeting with our partners to discuss the launch of a pilot using the es-coach. I was wondering if you have something that I could share with the team or whether you can give a tentative date when the team can see a demo? Thank you.
Alberto

**Comment by ultimike** [comment]
@escoach I just flew home yesterday - I'll touch base with Martin today and see how much progress he's made on the interface.

-mike

**Comment by ultimike** [comment]
@escoach Martin and I reviewed our progress a few minutes ago and the UI currently allows for the What/Who/Where to be selected and then the raw results are shown a JSON output (which we will use to generate the graphs and tables).

The following has not yet been implemented:

1. Data frame options (me, team, all)
2. Logic to prevent option selections we don't want the user to select (for example, a specific "who" and "where" at the same time).
3. Data table output.
4. Graph output.
5. Monthly/quarterly trends. 

We are happy to merge this into the `devel` branch so that you can play with it and possibly demo it on Thursday.

Additionally, @mgcobblestone will be working next on items 1 and 5 above, there is a possibility he can get one or both of these things implemented by end-of-day Wednesday.

-mike

**Comment by escoach** [comment]
@ultimike Thank you, I appreciate you working on this.

**Comment by ultimike** [comment]
@escoach We're currently working through some issues in an effort to provide you with something to demo tomorrow during your meeting.

What time is the meeting?

I'm also waiting on an (email) response from @thinkcollege regarding the development server (see #30).

-mike

**Comment by escoach** [comment]
@ultimike @thinkcollege the meeting is at 1 pm ET

**Comment by thinkcollege** [comment]
Alberto, I have the meeting at 11 on my calendar.

**Comment by escoach** [comment]
@thinkcollege I meant the meeting with the MN partners at 1pm.

**Comment by thinkcollege** [comment]
Oh, right, our internal meeting is on Thursday at 11.  I don't think you want to show the Minnesota people any of this stuff in its current unpolished state.  Maybe just describe it or show the screenshots.

**Comment by ultimike** [comment]
Ok - so no need to push out something for a demo then?

-mike

**Comment by thinkcollege** [comment]
I would like to demo it internally before showing it to external clients.  Alberto, would that work--at our Thursday meeting?

**Comment by ultimike** [comment]
There's not much to see yet - we're still working out bugs to return the proper data. 

-mike

**Comment by thinkcollege** [comment]
Or at some other time, since tomorrow is Thursday.  The heat is getting to me.

**Comment by escoach** [comment]
@ultimike when is the earliest that you can share a demo? 
@thinkcollege

**Comment by ultimike** [comment]
It depends on what is considered a "demo". So far, we've been focusing solely on implementing the SQL queries - not the interface. 

So, we have a /dashboard page with the available options. When selections are made and the form is submitted, we are returning data to the page in the form of JSON. We are currently testing/debugging this part. @mgcobblestone and I both consider this the most difficult part of the entire project. 

-mike

**Comment by thinkcollege** [comment]
I have no need to see a demo until this is further along.  @escoach and I are meeting this morning and we will determine what our timeline is for a demo and finished product.

**Comment by ultimike** [comment]
@mgcobblestone I did some limited testing on this late last week and all looks good so far. It could probably use some additional testing, if you have some time...

thanks,  
-mike

**Comment by ultimike** [comment]
It's time to play whack-a-mole!!!

Something broke both quarterly and monthly trends. ☹️ 

There's a PHP Notice in the log - this isn't the cause, but might as well fix while you're in there:

`Notice: Undefined index: TotalProvider in Drupal\survey_dashboard\Service\QueryBuilder->calculateHrs() (line 395 of /home/ici/escoach.communityinclusion.org/escoach/web/modules/custom/survey_dashboard/src/Service/QueryBuilder.php)`

There's also a PHP error - this could be part of the issue: `TypeError: Argument 1 passed to Drupal\survey_dashboard\Query\What::addSelectedWhatSums() must be of the type array, null given, called in /home/ici/escoach.communityinclusion.org/escoach/web/modules/custom/survey_dashboard/src/Service/QueryBuilder.php on line 477 in Drupal\survey_dashboard\Query\What->addSelectedWhatSums() (line 19 of /home/ici/escoach.communityinclusion.org/escoach/web/modules/custom/survey_dashboard/src/Query/What.php)`

-mike

**Comment by mgcobblestone** [comment]
@ultimike Trends seem to be working for me.  The error suggest that you may not have selected any "Whats", but there is validation in place that should throw an alert if that was the case.

Can you provide the details of the options you selected?  thx.

I did push up a fix for the warning.

**Comment by ultimike** [comment]
Hey @mgcobblestone, on [the dev site](http://sp-escoach.communityinclusion.org/dashboard), when I'm logged in as myself (part of the "no provider" team), I get no data on either Quarterly or Monthly trends.

I expect that "My" data columns will be 0, but not the "all" or "team" data.

Here's what I have selected:

- What: Getting to know a job seeker
- Who: (no selection)
- Where: (no selection)
- Timeframe: Quarterly (or Monthly)

I also tried changing my provider to "Institute for Community Inclusion" - same results.

-mike
